File: sudoq-app/sudoqmodel/src/main/java/de/sudoq/model/sudoku/sudokuTypes/ComplexityConstraintBuilder.java
package de.sudoq.model.sudoku.sudokuTypes;

import java.util.HashMap;
import java.util.Map;

import de.sudoq.model.sudoku.complexity.Complexity;
import de.sudoq.model.sudoku.complexity.ComplexityConstraint;
import de.sudoq.model.xml.XmlTree;
import de.sudoq.model.xml.Xmlable;

/**
 * Builder for Complexity constraints.
 */
public class ComplexityConstraintBuilder implements Xmlable {

	public Map<Complexity, ComplexityConstraint> specimen = new HashMap<>();
	
	public ComplexityConstraintBuilder(){
		
	}

	public ComplexityConstraintBuilder(Map<Complexity, ComplexityConstraint> specimen){
		this.specimen = specimen;
	}
	
	public ComplexityConstraint getComplexityConstraint(Complexity complexity){
		if (complexity != null && specimen.containsKey(complexity)) {
			return (ComplexityConstraint) specimen.get(complexity).clone();
		}
		return null;
	}
	
	
	public static final String TITLE = "ComplexityConstraintBuilder";
	
	@Override
	public XmlTree toXmlTree() {
		XmlTree representation = new XmlTree(TITLE);
		for (ComplexityConstraint v : specimen.values()){
			representation.addChild( v.toXmlTree() );
		}
		return representation;
	}

	@Override
	public void fillFromXml(XmlTree xmlTreeRepresentation) throws IllegalArgumentException {
		specimen = new HashMap<>();
		for (XmlTree sub : xmlTreeRepresentation) {
			ComplexityConstraint cc = new ComplexityConstraint();
			cc.fillFromXml(sub);
			specimen.put(cc.getComplexity(), cc);
		}

	}

}

File: sudoq-app/sudoqmodel/src/main/kotlin/de/sudoq/model/sudoku/sudokuTypes/ComplexityConstraintBuilder.kt
package de.sudoq.model.sudoku.sudokuTypes

import de.sudoq.model.sudoku.complexity.Complexity
import de.sudoq.model.sudoku.complexity.ComplexityConstraint
import java.util.*

/**
 * Builder for Complexity constraints.
 * todo vielleicht war das mal ein builder, aber jetzt wirkt es nutzlos, prüfen ob er entfernt werden kann
 */
class ComplexityConstraintBuilder(//todo this is not a "builder", rather a "holder", add assertions
    val specimen: Map<Complexity, ComplexityConstraint>
) {

    fun getComplexityConstraint(complexity: Complexity): ComplexityConstraint? {
        return if (specimen.containsKey(complexity)) {
            specimen[complexity]!!.clone() as ComplexityConstraint
        } else null
    }

}